### ISSUE TITLE
[Discuss][C++] Replace MemoTable with a SwissTable implementation

### ISSUE THREAD

**Issue by js8544** [apache/arrow#38372]
### Describe the enhancement requested

This is a follow up of https://github.com/apache/arrow/issues/36059#issuecomment-1592037527. There are many cases where we use a `MemoTable`, e.g. set lookup functions, vector hash functions, the `count_distinct` aggregate function, dictionary unification etc. Their performance can be boosted with a SwissTable.

We can either:
1. Use an existing swiss table library. This requires some work on dependency management. I recommend `absl::flat_hash_map` since they are the original author of swiss tables and we already has `absl` in our 3rd party toolchain.
2. Write one by ourselves. The current one in `acero` is too customized for the join node and it seems hard to extract a general hashtable from it. If I were to write one I would probably follow the structure of `absl`'s and replace things like memory management and bit tweaking with Arrow ones.

What do you think? @pitrou @westonpace 

### Component(s)

C++

**Comment by pitrou** [comment]
I don't know what a "Swiss table" is. Can you point to a general description?

**Comment by pitrou** [comment]
More generally, I'm all for a faster associative table implementation.

**Comment by mapleFU** [comment]
Maybe we should first having an memory-usage / performance benchmark on this?

**Comment by js8544** [comment]
There's an [Abseil's article](https://abseil.io/about/design/swisstables) and a [CppCon talk](https://www.youtube.com/watch?v=ncHmEUmJZf4) that provide a good intro. The key idea is, when looking up a key via linear probing, it can check multiple slots at once.

**Comment by pitrou** [comment]
Ah, so it's a hash table with SIMD-optimized lookup?

**Comment by js8544** [comment]
> Maybe we should first having an memory-usage / performance benchmark on this?

Okay, I'll try to draft a proof of concept with the `is_in` kernel.

**Comment by pitrou** [comment]
If you're going to use SIMD lookups, then I would ask that it relies on xsimd instead of intrinsics.

**Comment by js8544** [comment]
> Ah, so it's a hash table with SIMD-optimized lookup?

Well it's a speedup even without SIMD instructions. To put it in a oversimplified term each slot occupies 1 byte so with int64 we can already check 8 slots at the same time. But SIMD can definitely make it even faster.

> If you're going to use SIMD lookups, then I would ask that it relies on xsimd instead of intrinsics.

Sure, if we decide to write our own version.

**Comment by pitrou** [comment]
> Well it's a speedup even without SIMD instructions. To put it in a oversimplified term each slot occupies 1 byte so with int64 we can already check 8 slots at the same time. But SIMD can definitely make it even faster.

I see, thanks for the explanation.

> Sure, if we decide to write our own version.

Well, we don't want to depend on Abseil, so our own version it should probably be :-)

**Comment by llama90** [comment]
> There's an [Abseil's article](https://abseil.io/about/design/swisstables) and a [CppCon talk](https://www.youtube.com/watch?v=ncHmEUmJZf4) that provide a good intro. The key idea is, when looking up a key via linear probing, it can check multiple slots at once.

Thank you for explaining. I was also curious about what Swiss Join is. :)

**Comment by pitrou** [comment]
Can you post benchmark numbers on string values? While hashing integers is nice, I'm not sure that's a dominant use case.

**Comment by js8544** [comment]
Sure. I chose Int32 because it was the only one tested with a large value set. I just wrote a similar set of benchmarks for strings and here's the result.

### Before
<img width="1216" alt="image" src="https://github.com/apache/arrow/assets/14072670/ce751daf-2ec3-405e-98c4-6ee16401d7b7">

### After
<img width="1246" alt="image" src="https://github.com/apache/arrow/assets/14072670/6e992bcb-eb4d-41c9-8eae-fda54a0a7927">

**Comment by pitrou** [comment]
Ok. I agree we could replace MemoTable with something better. However, we don't want to have a dependency on Abseil, so it will have to be reimplemented.

**Comment by js8544** [comment]
> Ok. I agree we could replace MemoTable with something better. However, we don't want to have a dependency on Abseil, so it will have to be reimplemented.

Yes. I'll implement one using `xsimd` as you suggested. Also the swiss table may be worse than `MemoTable` on super small tables, since it checks 8 slots at once and reserves at least 8 slots on construction, as shown in the first two string benchmarks. We'll need further benchmarking to decide which hashtable to use in each individual use cases, or maybe with a dynamic dispatching mechanism.

**Comment by mapleFU** [comment]
I guess this might benifit the parquet dictionary encoding, which is widely used in parquet. If nobody move forward, I'll have a try next month

**Comment by js8544** [comment]
Yeah I'm afraid I won't have time to do this recently. Please have a try if you wish. Thanks!

**Comment by zanmato1984** [comment]
Will the swiss table currently being used in acero's hash join (aka. swiss join) help on this?

**Comment by js8544** [comment]
The swiss table for hash join is very specialized and kind of hard to generalize. It can't be directly used for memotable.

**Comment by SGZW** [comment]
Could I try to solve this issue?

**Comment by pitrou** [comment]
@SGZW Anyone can work on an issue without asking for permission. Feel free to experiment and submit a PR when ready.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.

**Comment by SGZW** [comment]
> This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.

i will finish this issue soon
